File: src/test/java/org/snakeyaml/engine/usecases/custom_collections/CustomDefaultCollectionsTest.java
/*
 * Copyright (c) 2018, SnakeYAML
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.snakeyaml.engine.usecases.custom_collections;

import static org.junit.jupiter.api.Assertions.assertEquals;

import java.util.LinkedList;
import java.util.TreeMap;
import java.util.TreeSet;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.snakeyaml.engine.v2.api.Load;
import org.snakeyaml.engine.v2.api.LoadSettings;

@org.junit.jupiter.api.Tag("fast")
class CustomDefaultCollectionsTest {

  @Test
  @DisplayName("Create LinkedList by default")
  void createLinkedListByDefault() {
    //init size is not used in LinkedList
    LoadSettings settings = LoadSettings.builder().setDefaultList(initSize -> new LinkedList())
        .build();
    Load load = new Load(settings);
    LinkedList<String> list = (LinkedList<String>) load.loadFromString("- a\n- b");
    assertEquals(2, list.size());
  }

  @Test
  @DisplayName("Create TreeMap by default")
  void createTreeMapByDefault() {
    //init size is not used in TreeMap
    LoadSettings settings = LoadSettings.builder().setDefaultMap(initSize -> new TreeMap()).build();
    Load load = new Load(settings);
    TreeMap<String, String> map = (TreeMap<String, String>) load.loadFromString("{k1: v1, k2: v2}");
    assertEquals(2, map.size());
  }

  @Test
  @DisplayName("Create TreeSet by default")
  void createTreeSetByDefault() {
    LoadSettings settings = LoadSettings.builder().setDefaultSet(initSize -> new TreeSet()).build();
    Load load = new Load(settings);
    TreeSet<String> set = (TreeSet<String>) load.loadFromString("!!set\n? foo\n? bar");
    assertEquals(2, set.size());
    //must be re-ordered
    assertEquals("bar", set.first());
    assertEquals("foo", set.last());
  }
}

File: src/commonTest/kotlin/it/krzeminski/snakeyaml/engine/kmp/usecases/custom_collections/CustomDefaultCollectionsTest.kt
@file:Suppress("UNCHECKED_CAST")
package it.krzeminski.snakeyaml.engine.kmp.usecases.custom_collections

import io.kotest.core.spec.style.FunSpec
import io.kotest.matchers.collections.shouldHaveSize
import io.kotest.matchers.maps.shouldHaveSize
import io.kotest.matchers.shouldBe
import io.kotest.matchers.types.shouldBeInstanceOf
import it.krzeminski.snakeyaml.engine.kmp.api.Load
import it.krzeminski.snakeyaml.engine.kmp.api.LoadSettings

class CustomDefaultCollectionsTest : FunSpec({
    test("create LinkedList by default") {
        // init size is not used in LinkedList
        val settings = LoadSettings(defaultList = { mutableListOf() })
        val load = Load(settings)
        val list = load.loadOne("- a\n- b") as List<String>
        list shouldHaveSize 2
    }

    test("create TreeMap by default") {
        // init size is not used in TreeMap
        val settings = LoadSettings(defaultMap = { mutableMapOf() })
        val load = Load(settings)
        val map = load.loadOne("{k1: v1, k2: v2}")
        map.shouldBeInstanceOf<Map<String, String>>()
        map shouldHaveSize 2
    }

    test("create TreeSet by default") {
        val settings = LoadSettings(defaultSet = { mutableSetOf() })
        val load = Load(settings)
        val set = load.loadOne("!!set\n? foo\n? bar")
        set.shouldBeInstanceOf<Set<String>>()
        set shouldHaveSize 2
        // must be re-ordered
        val sortedSet = set.sorted()
        sortedSet.first() shouldBe "bar"
        sortedSet.last() shouldBe "foo"
    }
})
